bam_qc_metrics/bam_qc.py

```diff
@@ -273,56 +288,156 @@ def read_mark_dup(self, input_path):
+            if aggregate_metrics['UNPAIRED_READS_EXAMINED'] + aggregate_metrics['READ_PAIRS_EXAMINED'] != 0:
+                aggregate_metrics['PERCENT_DUPLICATION'] = round(
+                    (aggregate_metrics['UNPAIRED_READ_DUPLICATES'] + aggregate_metrics['READ_PAIR_DUPLICATES'] * 2)
+                    /
+                    (aggregate_metrics['UNPAIRED_READS_EXAMINED'] + aggregate_metrics['READ_PAIRS_EXAMINED'] * 2), 5)
             else:
-                metrics[keys[i]] = int(values[i])
+                aggregate_metrics['PERCENT_DUPLICATION'] = 0.0
+
+            aggregate_metrics['ESTIMATED_LIBRARY_SIZE'] = math.trunc(self.estimate_library_size(
+                aggregate_metrics['READ_PAIRS_EXAMINED'] - aggregate_metrics['READ_PAIR_OPTICAL_DUPLICATES'],
+                aggregate_metrics['READ_PAIRS_EXAMINED'] - aggregate_metrics['READ_PAIR_DUPLICATES']))
+
+            aggregate_metrics['LIBRARY'] = 'AGGREGATE'
+
+            metrics.update(aggregate_metrics)
+            metrics['LIBRARY_METRICS'] = library_metrics
+
         if hist:
             metrics['HISTOGRAM'] = hist
+
         return metrics
 
+    @staticmethod
+    def estimate_library_size(read_pairs, unique_read_pairs):
+        """
+        NOTE: This function is directly based off of Picards estimateLibrarySize method
+        See: https://github.com/broadinstitute/picard/blob/78ea24466d9bcab93db89f22e6a6bf64d0ad7782/src/main/java/picard/sam/DuplicationMetrics.java#L135
+
+        Estimates the size of a library based on the number of paired end molecules observed
+        and the number of unique pairs observed.
+
+        Based on the Lander-Waterman equation that states:
+        C/X = 1 - exp( -N/X )
+        where
+        X = number of distinct molecules in library
+        N = number of read pairs
+        C = number of distinct fragments observed in read pairs
+        """
+
+        # Method that is used in the computation of estimated library size
+        f = lambda x, c, n: c / x - 1 + math.exp(-n / x)
+
+        read_pair_duplicates = read_pairs - unique_read_pairs
+        if read_pairs > 0 and read_pair_duplicates > 0:
+            m = 1.0
+            M = 100.0
+
+            if unique_read_pairs >= read_pairs or f(m * unique_read_pairs, unique_read_pairs, read_pairs) < 0:
+                raise Exception("Invalid values for pairs and unique pairs: " + read_pairs + ", " + unique_read_pairs)
+
+            # find value of M, large enough to act as other side for bisection method
+            while f(M * unique_read_pairs, unique_read_pairs, read_pairs) > 0:
+                M *= 10.0
+
+            # use bisection method (no more than 40 times) to find solution
+            for i in range(40):
+                r = (m + M) / 2.0;
+                u = f(r * unique_read_pairs, unique_read_pairs, read_pairs);
+                if u == 0:
```

Comment:
`u` is unlikely to be exactly zero, within floating point accuracy.
We could have a constant delta (set to a small positive value, eg. 0.000001) and have:

```
if math.fabs(u) < delta:
    break
elif u > 0:
    m = r
elif u < 0:
    M = r
```

Reply:
Good call, added check against machine epsilon